lib/modules/datasource/julia-pkg-server/registry.ts

```diff
@@ -0,0 +1,246 @@
+ * `<PkgServer host>/registry/<registry UUID>/<registry state>`.
+ *
+ * Although required when querying the state of a registry, the state in the
+ * URL is not mandatory when configuring a `registryUrl` passed into the
+ * datasource. Therefore that section is marked optional and should be ensured
+ * to be filled before further processing.
+ */
+const PKG_SERVER_REGISTRY_URL_FORMAT = new RegExp(
+  '^(?<pkgServer>.+)/registry/(?<uuid>\\p{Hex_Digit}{8}(-\\p{Hex_Digit}{4}){3}-\\p{Hex_Digit}{12})(/(?<state>\\p{Hex_Digit}{40}))?/?$',
+  'u',
+);
+
+/**
+ * The final part of the path of a (fully specified) registry URL references
+ * the tree SHA of the state of of the registry repository the package server
+ * is serving. Given the content-addressable nature of this SHA, it can be used
+ * as a caching mechanism.
+ *
+ * Cache keys are of the form `<registry UUID>:<registry state>`
+ */
+export function cacheKeyFromRegistryUrl(url: string): string {
+  return url.split('/').slice(-2).join(':');
+}
+
+/**
+ * Extracts one or more files from the provided tarball in-memory.
+ *
+ * Returns `null`/fails if not all requested files can be extracted or if no
+ * files are requested for extraction.
+ */
+export async function extractFilesFromTarball(
+  tarball: Buffer,
+  pathsToExtract: string[],
+): Promise<FileContents<string> | null> {
+  if (pathsToExtract.length === 0) {
+    return null;
+  }
+
+  // Block until the content of all files has been (asynchronously) read from
+  // the tarball
+  return await new Promise<FileContents<string> | null>((resolve, reject) => {
+    // File content may be streamed in chunks which need to be aggregated
+    // TODO: The tests for this functionality only require a single "chunk" to
+    // be read from the tarball to succeed. Hence, it would be sufficient to
+    // just process a single chunk directly as a string and return it. This
+    // does not hold in general and a (minimal) test requiring multiple chunks
+    // should be implemented
+    const fileContentChunks: FileContents<Buffer[]> = {};
+    for (const path of pathsToExtract) {
+      fileContentChunks[path] = [];
+    }
+
+    const tarballParser = new tarParse({
+      filter: (path) => pathsToExtract.includes(path),
+    });
+
+    tarballParser.on('end', () => {
+      if (Object.values(fileContentChunks).every(({ length }) => length > 0)) {
+        const fileContents: FileContents<string> = {};
+        for (const path of pathsToExtract) {
+          fileContents[path] = Buffer.concat(fileContentChunks[path]).toString(
+            'utf-8',
+          );
+        }
+
+        resolve(fileContents);
+      } else {
+        resolve(null);
+      }
+    });
+
+    tarballParser.on('entry', (entry) => {
+      entry.on('data', (data) => {
+        fileContentChunks[entry.path].push(data);
+      });
+
+      entry.on('error', reject);
+    });
+
+    tarballParser.end(tarball);
```

Comment:
This extracts the necessary files in memory. It might be more straightforward to unpack do disk instead and work with some form of caching. I'd be happy to change it if that's better suited to Renovate's overall design.

Reply:
it depends on how big these tarballs can be. ~10 MB should be ok, but bigger files can cause memory issues (oom)

Reply:
@rarkins we need to think about those big blobs. we cache every request in memory, so we've this blob twice.

https://github.com/renovatebot/renovate/blob/26337ac124f4dd1885a44ed65822f2944ad69f3d/lib/util/http/index.ts#L186

we probably should pass `memCache:false` for those files and only cache the smaller transformed data to package cache.

@bauglir I think we should extract all deps from that file and add them to the package cache, so we can speed up additional lookups. we do something similar at `rubygems` datasource.